### PR TITLE
PSM Data Feed Support, DockerHub Docs, Commsignia Log Offloading Support

### PR DESCRIPTION
**Pull request by dmccoystephenson**
# PR Details
### PSM Data Feed Support
This PR introduces support for PSM message processing to the ODE. Support for both required and optional fields is present, however a sample message that covers any of the optional fields is not available for testing. All required fields have been tested with sample PSM messages generated from the NJDOT deployment.

This pipeline takes in PSM messages on the new UDP port (44940) and outputs processed JSON PSM messages on the topic: topic.OdePsmJson. As with other data feeds, multiple other topics are used for internal processing/management: topic.OdeRawEncodedPSMJson and topic.OdePsmTxPojo.

### DockerHub Documentation
Documentation for using the image has been added to `docs/dockerhub.md`.

### Commsignia Log Offloading Support
This PR adds support for processing log offloading from Commsignia devices. The primary change here is supporting the new line characters that are included in these log messages, as well as parsing off the WSMP message header in the payload.

### Additional Changes
- Deprecated some test messages
- Added new logging variables for PPM/ACM to docker-compose.yml & sample.env files

### Commsignia Log Offloading Support
To process offloaded logs in both CDOT/WYDOT deployment. In preparation of the anticipated firmware release that supports log offloading.

### PSM Data Feed Support
Required for processing Personal Safety Messages through the ODE. 

### PSM Data Feed Support
The convention from previous data feeds was followed for unit testing and verified message processing with sample messages from the NJDOT deployment.

### Commsignia Log Offloading Support
Added new unit tests to verify log processing of old messages (without newline characters).
Tested the log offloading of all .gz example log files in the data directory. To run this test, run the ODE through docker, at the minimum run the zookeeper, kafka, ODE, AEM, and ADM services. Copy all of the .gz files from the jpo-ode/data folder into the jpo-ode/uploads/bsmlog folder. This folder is monitored by the ODE and the files are ingested into the appropriate topics. If the processing succeeds the files are forwarded to the jpo-ode/uploads/backup folder and if they fail it goes to the jpo-ode/uploads/failed directory.

## Types of changes
- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)
- [x] New Documentation

## Checklist:
- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/ImporterDirectoryWatcher.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
 public class ImporterDirectoryWatcher implements Runnable {
    
    public enum ImporterFileType {
-      LEAR_LOG_FILE, JSON_FILE
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/ImporterDirectoryWatcher.java:32]: What is the diff between JSON and LOG file?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 2]
```diff
@@ -186,4 +191,43 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
+		} catch (JsonUtilsException e) {
+			logger.error("JsonUtilsException while checking message header. Stacktrace: " + e.toString());
+
+		}
+		return header;
+	}
+
+	// This method will check if the next character is a newline character (0x0A in hex or 10 in converted decimal) 
+	// or if the next character does not contain a newline character it will put that character back into the buffered input stream
+	public BufferedInputStream removeNextNewLineCharacter(BufferedInputStream bis) {
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:221]: Do we have a unit test for this?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 3]
```diff
@@ -186,4 +191,43 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
+				logger.debug("Payload JSON: " + payloadJson);
+			}
+		} catch (JsonUtilsException e) {
+			logger.error("JsonUtilsException while checking message header. Stacktrace: " + e.toString());
+
+		}
+		return header;
+	}
+
+	// This method will check if the next character is a newline character (0x0A in hex or 10 in converted decimal)
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:219]: Complete the if there is a newline character statement?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 4]
```diff
@@ -20,18 +20,31 @@
...
+   private static final int HEADER_SIZE = 20; 
+   public static final int PAYLOAD_LENGTH = 2;
+   
    protected short payloadLength;
    protected byte[] payload;
+   protected String payloadType;
 
    public PayloadParser() {
       super();
+      msgStartFlags.put("BSM", "0014");
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:45]: Constant variables for the flag values?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 5]
```diff
@@ -90,4 +105,37 @@ public void writeTo(OutputStream os) throws IOException {
...
 
+
+
+   public byte[] removeHeader(byte[] packet) {
+      String hexPacket = HexUtils.toHexString(packet);
+      String hexPacketParsed = "";
+      for (String key : msgStartFlags.keySet()) {
+         String startFlag = msgStartFlags.get(key);
+         int startIndex = hexPacket.toLowerCase().indexOf(startFlag);
+         if (hexPacketParsed == ""){
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:116]: Maybe use equals()?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 6]
```diff
@@ -90,4 +105,37 @@ public void writeTo(OutputStream os) throws IOException {
...
+            // Using a value of 35 as the largest index from preliminary testing data.
+            } else if (startIndex > HEADER_SIZE && startIndex < 35) {
+               int trueStartIndex = HEADER_SIZE
+                     + hexPacket.substring(HEADER_SIZE, hexPacket.length()).indexOf(startFlag);
+               logger.debug("Found payload start at: " + trueStartIndex);
+               hexPacketParsed = hexPacket.substring(trueStartIndex, hexPacket.length());
+            }
+         }
+      }
+      if (hexPacketParsed == ""){
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:133]: Maybe use equals()?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 7]
```diff
@@ -90,4 +105,37 @@ public void writeTo(OutputStream os) throws IOException {
...
+         if (hexPacketParsed == ""){
+            logger.debug("Start index for: " + key + " is: " + startIndex);
+            if (startIndex == -1) {
+               logger.debug("Message does not have header for: " + key);
+               break;
+            } else if (startIndex <= HEADER_SIZE) {
+               logger.debug("Message has supported header. startIndex: " + startIndex + " msgFlag: " + startFlag);
+               hexPacketParsed = hexPacket;
+            // Using a value of 35 as the largest index from preliminary testing data.
+            } else if (startIndex > HEADER_SIZE && startIndex < 35) {
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:125]: Maybe make 35 consistent with HEADER_SIZE?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 8]
```diff
@@ -69,6 +74,9 @@ public LogFileToAsn1CodecPublisherException(String string, Exception e) {
 	public LogFileToAsn1CodecPublisher(StringPublisher dataPub) {
 		this.publisher = dataPub;
 		this.serialId = new SerialId();
+		msgStartFlags.put("BSM", "0014");
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:77]: Do we need a start flag for PSM too?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java (hunk 4 above)
Comment (dan-du-car) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:45]: Do we need one for PSM too?

**Comment by dan-du-car** [comment]
```
git clone --recurse-submodules https://github.com/CDOT-CV/jpo-ode
cd jpo-ode
mvn test
```
image

**Comment by dmccoystephenson** [comment]
> ```
> git clone --recurse-submodules https://github.com/CDOT-CV/jpo-ode
> cd jpo-ode
> mvn test
> ```
> 
> image

The proposed changes are on the 'dev' branch. I recommend using the following commands to test:
```
git clone --recurse-submodules https://github.com/CDOT-CV/jpo-ode
cd jpo-ode
git checkout origin/dev
mvn test
```

**Comment by dmccoystephenson** [comment]
What output do you get when running `mvn --version`? @dan-du-car

Do you experience the same behavior when using the dev container for the project?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:77]: A start flag for PSM messages is not necessary currently, as we don't currently support offloaded PSM messages through log files (it is also not supported by Commsignia/Yunex devices).

**Comment by Michael7371** [comment]
> > > ```
> > > git clone --recurse-submodules https://github.com/CDOT-CV/jpo-ode
> > > cd jpo-ode
> > > mvn test
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > image
> > 
> > 
> > The proposed changes are on the 'dev' branch. I recommend using the following commands to test:
> > ```
> > git clone --recurse-submodules https://github.com/CDOT-CV/jpo-ode
> > cd jpo-ode
> > git checkout origin/dev
> > mvn test
> > ```
> 
> Yes, it was in dev. image

Have you followed the "Downloading the source code from a non-default branch" section of the root readme? It may help with getting the submodule references setup the same way as the dev branch does.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 9]
```diff
@@ -146,38 +155,34 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
+			} else {
 				if (isBsmRecord()) {
 					logger.debug("Publishing a BSM");
-				} else if(isSpatRecord()) {
+				} else if (isSpatRecord()) {
 					logger.debug("Publishing a Spat");
-				}else {
-					logger.debug("Publishing a TIM");
+				} else {
+					logger.debug("Publishing a TIM or MAP");
```
[the remark this answers is not in the document]
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:172]: We would need to add more logic to determine what is a TIM/MAP message from the RxMsgFileParser Class. Doesn't seem necessary for this PR but could be added later.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 10]
```diff
@@ -186,4 +191,43 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
+		String header = null;
+		try {
+			payloadJson = JsonUtils.toJSONObject(payload.getData().toJson());
+			String hexPacket = payloadJson.getString("bytes");
+
+			for (String key : msgStartFlags.keySet()) {
+				String startFlag = msgStartFlags.get(key);
+				int startIndex = hexPacket.toLowerCase().indexOf(startFlag);
+				logger.debug("Start index for " + key + "(" + startFlag + ")" + " is: " + startIndex);
+				if (startIndex <= 20 && startIndex != 0 && startIndex != -1) {
```
[the remark this answers is not in the document]
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:205]: Yeah -1 is not found, 0 is that the payload start string is already at the start of the payload (no header), and greater than 20 is likely not associated with the start of the payload and just part of the payload itself. This logic's purpose is to remove headers that the ODE currently can't process/decode.
The logic originated from the UDP receiver [classes](https://github.com/CDOT-CV/jpo-ode/blob/dev/jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiver.java)

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 11]
```diff
@@ -186,4 +191,43 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
 		}
 	}
 
+	public String checkHeader(OdeMsgPayload payload) {
```
[the remark this answers is not in the document]
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:194]: Yeah, this logic determines if the message has a Ieee1609Dot2Data payload header on it. If the message has this header, we pass through that information to the ASN1 codec to allow it to parse it for data.
It doesn't modify the payload during this step.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/ImporterDirectoryWatcher.java (hunk 1 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/ImporterDirectoryWatcher.java:32]: The main difference is that JSON is uncompressed and the LOG file is a gzip binary file. I think the old DSRC Lear devices used to have this as a configurable setting. Current devices only support the gzip binary file output.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java (hunk 2 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:221]: We have an [integration test](https://github.com/CDOT-CV/jpo-ode/tree/dev/jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisherTest.java) in the test named: "testPublishRxMsgBSMLogFileNewLine" that verifies its functionality but not a specific unit test:

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java (hunk 3 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:219]: We don't need any logic to happen when there is a new line character, as the newline character is already read out of the buffer. If the next character isn't a new line character, we reset it to support log files that don't have new line characters.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java (hunk 4 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:45]: Same response as above:
A start flag for PSM messages is not necessary currently, as we don't currently support offloaded PSM messages through log files (it is also not supported by Commsignia/Yunex devices).

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java (hunk 7 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:125]: To add clarity, I will set the 35 value to a constant named: "HEADER_SIZE_WSMP", as this logic is intended to remove the unsupported WSMP payload header from the message.
We hope to support WSMP in the future, but we currently don't have the ability to process the header contents.

**Comment by dmccoystephenson** [comment]
This PR is ready for re-review.

**Review by dan-du-car — APPROVED** [review]

**Comment by dan-du-car** [comment]
Looks good. Thank you!

**Pull request merged**